README.md
# swgoh-telegram
Test #2, Now with telegram


changelog.txt
======SWGOH API bot's changeLog======

=====version 0.1.0 by 2018 Aug 06=====
* alpha-version of bot;
* Add command "/getallchars" to show list of characters in game;
* Add command /getchar:CHARNAME to show progress by this characters of Guild members;
* Add command "/guild" to show current guild bot is working with;

=====version 0.1.1 by 2018 Aug 08=====
* Add command "/getphase:PHASENUMBER to show players with full packs for this phase;

arrays.js
//Repository for data

//Packs of characters for phases
const packsForSiths = {
    phase1: [
        'REYJEDITRAINING',
        'REY',
        'R2D2_LEGENDARY',
        'BB8',
        'RESISTANCETROOPER'
    ],
    phase2: [
        'VEERS',
        'COLONELSTARCK',
        'SNOWTROOPER',
        'MAGMATROOPER',
        'SHORETROOPER',
        'STORMTROOPER',
        'GRANDADMIRALTHRAWN'
    ],
    phase3: [
        'COMMANDERLUKESKYWALKER',
        'HANSOLO',
        'PAO',
        'DEATHTROOPER',
        'CHIRRUTIMWE'
    ],
    phase4: [
        'ASAJVENTRESS',
        'DAKA',
        'MOTHERTALZIN',
        'NIGHTSISTERZOMBIE',
        'TALIA',
        'NIGHTSISTERINITIATE',
        'NIGHTSISTERACOLYTE'
    ]
};

//Replies for unknown commands to bot
const replies = [
    "Sorry, Master, I know languages of 6,000,000 life forms but I don't understand you! Please try again.",
    "I have a bad felling about this...",
    "It's a trap!",
    "Help me, Obi-Wan Kenobi. You’re my only hope.",
    "I find your lack of faith disturbing.",
    "Do. Or do not. There is no try.",
    "Pew-pew-pew",
    "I’m one with the Force. The Force is with me.",
    "May the force be with you.",
    "Use the force, Luke.",
    "Fear is the path to the dark side.",
    "It was said that you would destroy the Sith, not join them."
  ];

  //Message for raids

  const raidMessage = (tankBeginDate, tankEndDate, pitBeginDate, pitEndDate, sithsBeginDate, sithsEndDate) => {
return `👉Расписание рейдов на Август!
  ✅ Все даты могут быть изменены! Следите за расписанием.
  
  🚧🚧🚧 ТАНК
  ✳️  ${tankBeginDate} (отметка на 0)
  
  ${tankEndDate}:
  Фаза 1 только для ДВ - 13:00 МСК.
  
  Фаза 2-4 - 20:00 МСК.
  
  🔺Урон свыше 10% закрывать минимум через 15 минут после начала фазы.
  Режим 1 фаза / 1 пачка.
  👮♀Наказания за нарушения:
  
  
  🐷🐷🐷 РАНКОР
  ✳️ ${pitBeginDate}
  
  🔺Соло закрытие - ${pitEndDate}. в 21.15 по МСК (ЗЕРГ)
  
  👉ДВ бьют ф1 ${pitEndDate} в 12:15 по Мск одновременно, закрывают попытку не раньше 12:30 по Мск. Максимальный перенесенный урон с ф1 на ф2 50к.
  Соответственно, в остальных фазах они не принимают участия.
  Kot @265290259
  Fesko
  Erast
  👉Все остальные начинают бить ${pitEndDate} в 21.00 Если ваш отряд(ы) умирает до 21:05, ваш зафиксированный урон в сумме не должен превышать 500к (16-17%)
  👮♀Наказания за нарушения:
   @vA_Dmi3y пропуск
  
  👹👹👹 СИТХИ
  ✳️ ${sithsBeginDate}
  
  🛑Этап 1 (1-2 фаза) - в 19.00 МСК (1 фаза –Рей, 2 фаза – 3 любых профильных пака)
  
  👉Следующий день (${sithsEndDate})
  🛑Этап 2 (3 фаза) - в 11.00 мск
                    (4 фаза) - в 19.30 мск
  
  👮Наказания за нарушения:`;
  };

  module.exports.raidMessage = raidMessage;
  module.exports.replies = replies;
  module.exports.packsForSiths = packsForSiths;

getchars.js
//Module for listing progress of character

const fetch = require('node-fetch');

const getCurrentChar = async (url, char) => {
    const getData = await fetch(url);
    console.log(getData.status);
    const data = await getData.text();
    const dataToJson = JSON.parse(data);
    const getKeys = Object.keys(dataToJson);
    if (!char) {
        return 'Enter character name, please! Format: getchar:CHARNAME'
    }
    if (char === 'getallchars') {
        //Return list of all characters
        return getKeys.join('\n');
    }
    else {
        const findChar = getKeys.filter(element => element === char);
        if (findChar.length === 0) {
            return 'No characters found!'
        }
        const currentChar = dataToJson[findChar];
        const makeView = currentChar.map(element => {
            const { player, gear_level, level, rarity } = element;
            return [`Player: ${player}`, ` stars: ${rarity} `, ` gear level: ${gear_level}`];
        });
        console.log('Character is ready');
        return makeView.join('\n');
    }
  };

module.exports.getCurrentChar = getCurrentChar;

filterSiths.js
//Module to calculate user who are ready for every phase

require('dotenv').config();

const fetch = require('node-fetch');

const arrays = require('./arrays.js');

var guildId = process.env.DEFAULT_ID;

const getReadyForPhase = async (url, phase) => {
    const getData = await fetch(url);
    console.log(getData.status);
    const data = await getData.text();
    const dataToJson = JSON.parse(data);
    const arrayPhase = arrays.packsForSiths[phase];
    //Filter data to select only 7-stars characters in phase's array and sort by user
    const findFullPack = arrayPhase.reduce((acc, element) => {
        const getPlayers = dataToJson[element].reduce((counter, entry) => {
            if (entry['rarity'] === 7) {
                const { player, gear_level } = entry;
                const findPlayer = Object.keys(acc).filter(user => user === entry['player']);
                if (findPlayer.length === 0) {
                    return { ...counter, [player]: { [element]: gear_level } };
                }
                counter[player] = {...acc[player], [element]: gear_level };
                return counter;
            }
            return counter;
        }, acc);
        return getPlayers;
    }, {});
    //Filter result to select users with mininum 5 characters in pack.
    const filterWhoReady = {};
    for (const record in findFullPack) {
        if (Object.keys(findFullPack[record]).length >= 5) {
            filterWhoReady[record] = findFullPack[record];
        }
    }
    const totalReady = Object.keys(filterWhoReady).length;
    console.log(totalReady);
    //Change view to print for bot's user.
    const makeViewSets = Object.keys(filterWhoReady).map(element => {
            const makePlayer = `${element}: `;
            const makeSet = Object.keys(filterWhoReady[element]).map(char => `${char}: ${filterWhoReady[element][char]}`).join(', ');
            return [makePlayer, makeSet].join('\n');
    });
    //Paging result to reduce size of data
    const page1 = makeViewSets.slice(0, makeViewSets.length / 2).join('\n');
    const page2 = makeViewSets.slice((makeViewSets.length / 2) + 1).join('\n');
    return [page1, page2, totalReady];
  };

module.exports.getReadyForPhase = getReadyForPhase;

app.js
//main module to rule the bot

require('dotenv').config();

//Importing other modules
const getReadyForPhase = require('./filterSiths.js');
const getCurrentChar = require('./getchars.js');
const arrays = require('./arrays.js');
const fs = require('fs');

const TelegramBot = require('node-telegram-bot-api');
const token = process.env.TOKEN;
const guildId = process.env.DEFAULT_ID;
const bot = new TelegramBot(token, {polling: true});

const urlParse = `https://swgoh.gg/api/guilds/${guildId}/units/`;
const firtsDate = new Date(2018, 7, 13);
const beginBank = []

//bot's commands
bot.on('message', function (msg) {
    let chatId = msg.chat.id;
    var charName = '';
    const userName = (msg.chat.username) ? ` ${msg.chat.username}` : '';
    console.log(msg);
    switch (true) {
        case (msg.text === '/start'):
            console.log('Greetings');
            bot.sendMessage(chatId, `Oh dear! Master${userName}, it's so good to see you fully functional again.`, {caption: "Hello!"});
            break;
        case (msg.text.includes('/getchar')):
            console.log('Prepare char');
            var charName = msg.text.split(':')[1];
            console.log(charName);
            getCurrentChar.getCurrentChar(urlParse, charName).then(res => bot.sendMessage(chatId, res, {caption: "It's work!"}));
            break;
        case (msg.text === '/getallchars'):
            var charName = 'getallchars';
            console.log('List of chars');
            getCurrentChar.getCurrentChar(urlParse, charName).then(res => bot.sendMessage(chatId, res, {caption: "It's work!"}));
            break;
        case (msg.text === '/guild'):
            console.log('My guild');
            const res = `Your guild ID is ${guildId}`;
            bot.sendMessage(chatId, res, {caption: "It's work!"});
            break;
        case(msg.text.includes('/getphase')):
        //Validate phase number
            const parsePhase = Number(msg.text.split(':')[1]);
            if (isNaN(parsePhase)) {
                console.log('Wrong declaration of phase');
                bot.sendMessage(chatId, 'This is not phase you are looking for!', {caption: "You do it wrong!"});
                break;
            }
            const phaseNum = Math.round(parsePhase);
            console.log(phaseNum);
            if (phaseNum < 1 || phaseNum > 4 ) {
                console.log('Wrong number of phase');
                bot.sendMessage(chatId, 'This is not phase you are looking for!', {caption: "You do it wrong!"});
                break;
            }
            console.log('Packs for Phase');
            getReadyForPhase.getReadyForPhase(urlParse, `phase${phaseNum}`).then(res => {
                bot.sendMessage(chatId, res[0], {caption: "It's work!"});
                bot.sendMessage(chatId, res[1], {caption: "It's work!"});
                bot.sendMessage(chatId, `TOTAL READY: ${res[2]}`, {caption: "It's work!"});
            });
            break;
        //Admin's section
        case(msg.text === '/penaltieslist' && msg.from.username === 'vA_Tuor4eg'):
            const getPeenalties = fs.readFileSync('/tmp/penalties.txt').toString();
            console.log('List of penalties is prepared.')
            const penaltiesMessage = (getPeenalties !== '') ? getPeenalties : 'Excellent news, Master! There is no penalties for now!';
            bot.sendMessage(chatId, penaltiesMessage, {caption: "It's work!"});
            break;
        case(msg.text === '/clearpenalties' && msg.from.username === 'vA_Tuor4eg'):
            fs.writeFile('/tmp/penalties.txt', '', () => {
                console.log('List is clear');
                bot.sendMessage(chatId, 'List is clear', {caption: "It's work!"});
            });
            break;
        case(msg.text.includes('/set') && msg.from.username === 'vA_Tuor4eg'):
            const [command, user, raid] = msg.text.split(':');
            console.log(user, raid);
            const writeRes = `${user}: ${raid}\n`;
            fs.appendFile('/tmp/penalties.txt',  writeRes, (err) => {
                if (err) throw err;
                console.log(writeRes);
                console.log('The penalties list has been udated!');
                bot.sendMessage(chatId, 'Update list', {caption: "It's work!"});
              });
            break;
        case(msg.text === '/makemsg' && msg.from.username === 'vA_Tuor4eg'):
            const currentDate = new Date();
            console.log(firtsDate, currentDate);
            bot.sendMessage(chatId, arrays.raidMessage('14 августа', '15 августа'), {caption: "It's work!"});
            break;
        //End of admin's section
        case(msg.text === '/history'):
            const getChangelog = fs.readFileSync('./changelog.txt').toString();
            console.log('Changelog is prepared.')
            bot.sendMessage(chatId, getChangelog, {caption: "It's work!"});
            break;
        case(msg.text === '/version'):
            const getVersion = fs.readFileSync('./version.txt').toString();
            console.log('Version of bot')
            bot.sendMessage(chatId, `The current version is ${getVersion}.`, {caption: "It's work!"});
            break;
        default:
            console.log('something wrong');
            const genRandom = Math.floor(Math.random() * arrays.replies.length);
            const result = arrays.replies[genRandom];
            bot.sendMessage(chatId, result, {caption: "You do it wrong!"});
            break;
    }
    if (msg.from.username) {
        const readyToWrite = JSON.stringify({ from: msg.from, guild: guildId });
        console.log(JSON.parse(readyToWrite));
    }

});



